### PR TITLE
Add a force transfrom meta from v2 to v3

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: ref #3594 

Problem Summary:
- After user running in mix mode and do the delta merge.
- Meta won't be `del`, and table won't change to the ONLY_V3

### What is changed and how it works?
- Force transform the `meta` from V2 to V3 when `storagepool` restored.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 1]
```diff
@@ -273,6 +324,16 @@ PageStorageRunMode StoragePool::restore()
         auto v2_data_max_ids = data_storage_v2->restore();
         auto v2_meta_max_ids = meta_storage_v2->restore();
 
+        if (auto meta_remain_pages = meta_storage_v2->getNumberOfPages(); meta_remain_pages != 0)
+        {
+            bool traslate_done = forceTransformMetaV2toV3();
```
Comment (jiaqizho) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:329]: transform_done

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 2]
```diff
@@ -273,6 +322,21 @@ PageStorageRunMode StoragePool::restore()
         auto v2_data_max_ids = data_storage_v2->restore();
         auto v2_meta_max_ids = meta_storage_v2->restore();
 
+        if (const auto & meta_remain_pages = meta_storage_v2->getNumberOfPages(); meta_remain_pages != 0)
+        {
+            forceTransformMetaV2toV3();
+            const auto & meta_remain_pages_after_transform = meta_storage_v2->getNumberOfPages();
+            LOG_FMT_INFO(logger, "Current meta translate to V3 finished. [ns_id={}] [done={}] [remain_before_translate_pages={}], [remain_after_translate_pages={}]", //
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:329]:
```suggestion
            LOG_FMT_INFO(logger, "Current meta transform to V3 finished. [ns_id={}] [done={}] [pages_before_transform={}], [pages_after_transform={}]", //
```

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 3]
```diff
@@ -273,6 +322,21 @@ PageStorageRunMode StoragePool::restore()
         auto v2_data_max_ids = data_storage_v2->restore();
         auto v2_meta_max_ids = meta_storage_v2->restore();
 
+        if (const auto & meta_remain_pages = meta_storage_v2->getNumberOfPages(); meta_remain_pages != 0)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:325]: add some comments about why

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 4]
```diff
@@ -273,6 +322,21 @@ PageStorageRunMode StoragePool::restore()
...
+        {
+            forceTransformMetaV2toV3();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:327]: add some logs to indicate we start the transform process

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 5]
```diff
@@ -237,6 +237,55 @@ StoragePool::StoragePool(Context & global_ctx, NamespaceId ns_id_, StoragePathPo
...
+    for (const auto & page_transform : pages_transform)
+    {
+        // Check pages have not contain field offset
+        // Also get the tag of page_id
+        const auto & page_transform_entry = meta_transform_storage_reader->getPageEntry(page_transform.page_id);
+        if (!page_transform_entry.field_offsets.empty())
+        {
+            throw Exception(fmt::format("Can't transfrom meta from V2 to V3, [page_id={}]",
+                                        page_transform.page_id),
+                            ErrorCodes::LOGICAL_ERROR);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:265]:
```suggestion
            throw Exception(fmt::format("Can't transfrom meta from V2 to V3, [page_id={}] [entry={}]",
                                        page_transform.page_id, toDebugString(page_transform_entry)),
                            ErrorCodes::LOGICAL_ERROR);
```

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 6]
```diff
@@ -273,6 +319,25 @@ PageStorageRunMode StoragePool::restore()
         auto v2_data_max_ids = data_storage_v2->restore();
         auto v2_meta_max_ids = meta_storage_v2->restore();
 
+        // Meta part can rewrite by `delta merge`.
+        // But some of page won't be delete in `delta merge`.
+        // Need use forceTransformMetaV2toV3 to convert from V2 to V3.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:324]:
```suggestion
        // The pages on data and log can be rewritten to V3 and the old pages on V2 are deleted by `delta merge`. However, the pages on meta V2 can not be deleted. As the pages in meta is small, we perform a  forceTransformMetaV2toV3 to convert pages before all.
```

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 7]
```diff
@@ -273,6 +319,23 @@ PageStorageRunMode StoragePool::restore()
         auto v2_data_max_ids = data_storage_v2->restore();
         auto v2_meta_max_ids = meta_storage_v2->restore();
 
+        // The pages on data and log can be rewritten to V3 and the old pages on V2 are deleted by `delta merge`. However, the pages on meta V2 can not be deleted. As the pages in meta is small, we perform a  forceTransformMetaV2toV3 to convert pages before all.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:322]:
```suggestion
        // The pages on data and log can be rewritten to V3 and the old pages on V2 are deleted by `delta merge`.
        // However, the pages on meta V2 can not be deleted. As the pages in meta are small, we perform a forceTransformMetaV2toV3 to convert pages before all.
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**